sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/ByteBuddyDoFnInvokerFactory.java

```diff
@@ -516,36 +439,53 @@ private static MethodDescription getExtraContextFactoryMethodDescription(
     }
   }
 
-  private static StackManipulation simpleExtraContextParameter(
-    String methodName,
-    StackManipulation pushExtraContextFactory) {
+  /**
+   * Calls a zero-parameter getter on the {@link ExtraContextFactory}, which must be on top of the
+   * stack.
+   */
+  private static StackManipulation simpleExtraContextParameter(String methodName) {
       return new StackManipulation.Compound(
-        pushExtraContextFactory,
         MethodInvocation.invoke(getExtraContextFactoryMethodDescription(methodName)));
   }
 
   static StackManipulation getExtraContextParameter(
       DoFnSignature.Parameter parameter,
-      final StackManipulation pushExtraContextFactory) {
+      final StackManipulation pushDelegate) {
 
     return parameter.match(
         new Cases<StackManipulation>() {
 
           @Override
+          public StackManipulation dispatch(ContextParameter p) {
+            return new StackManipulation.Compound(
+                pushDelegate,
+                MethodInvocation.invoke(
+                    getExtraContextFactoryMethodDescription("context", DoFn.class)));
+          }
+
+          @Override
+          public StackManipulation dispatch(ProcessContextParameter p) {
+            return new StackManipulation.Compound(
+                pushDelegate,
+                MethodInvocation.invoke(
+                    getExtraContextFactoryMethodDescription("processContext", DoFn.class)));
+          }
+
+          @Override
           public StackManipulation dispatch(WindowParameter p) {
             return new StackManipulation.Compound(
-                simpleExtraContextParameter("window", pushExtraContextFactory),
+                simpleExtraContextParameter("window"),
```

Comment:
@jkff I took your advice and always presume the `ExtraContextFactory` is on the top of the stack.